runtime/src/main/cpp/Memory.cpp

```diff
@@ -1548,52 +1618,143 @@ void UpdateHeapRefIfNull(ObjHeader** location, const ObjHeader* object) {
+  // For all frozen objects in stack slots - perform reference increment.
+  FrameOverlay* frame = state->currentFrame;
+  MEMORY_LOG("actualizeNewlyFrozenOnStack: newly frozen size is %d\n", newlyFrozen->size())
+  while (frame != nullptr) {
+    MEMORY_LOG("current frame %p: %d parameters %d locals\n", frame, frame->parameters, frame->count)
+    ObjHeader** current = reinterpret_cast<ObjHeader**>(frame + 1) + frame->parameters;
+    ObjHeader** end = current + frame->count - kFrameOverlaySlots - frame->parameters;
+    while (current < end) {
+      ObjHeader* obj = *current;
+      current++;
+      if (obj != nullptr) {
+        auto* container = obj->container();
+        // No need to use atomic increment yet, object is still local.
+        if (container != nullptr && container->frozen() && newlyFrozen->count(container) != 0) {
+          IncrementRC<false>(container);
+        }
+      }
+    }
+    frame = frame->previous;
+  }
+
+  // And actualize RC of those objects using toRelease set.
+  for (auto& container : *(state->toRelease)) {
+    if (!isMarkedAsRemoved(container) && container->frozen()) {
+      RuntimeAssert(newlyFrozen->count(container) != 0, "Must be newly frozen");
+      ContainerHeader* realContainer = reinterpret_cast<ObjHeader*>(container + 1)->container();
+      auto oldRc = realContainer->decRefCount<false>();
+      MEMORY_LOG("decremented rc of %p to %d\n", realContainer, oldRc);
+      container = markAsRemoved(container);
+    }
+  }
+}
+
+void processDecrements(MemoryState* state) {
+  auto* toRelease = state->toRelease;
+  state->gcSuspendCount++;
+  while (toRelease->size() > 0) {
+     auto* container = toRelease->back();
+     toRelease->pop_back();
+     if (isMarkedAsRemoved(container))
+       continue;
+     DecrementRC(container);
+  }
+  state->gcSuspendCount--;
+}
+
+void decrementStack(MemoryState* state) {
+  state->gcSuspendCount++;
+  FrameOverlay* frame = state->currentFrame;
+  while (frame != nullptr) {
+    ObjHeader** current = reinterpret_cast<ObjHeader**>(frame + 1) + frame->parameters;
+    ObjHeader** end = current + frame->count - kFrameOverlaySlots - frame->parameters;
+    while (current < end) {
+      ObjHeader* obj = *current++;
+      if (obj != nullptr) {
+        auto* container = obj->container();
+        if (container != nullptr && container->tag() == CONTAINER_TAG_NORMAL)
+          EnqueueDecrementRC</* CanCollect = */ false>(container);
+      }
+    }
+    frame = frame->previous;
+  }
+  state->gcSuspendCount--;
+}
+
 void GarbageCollect() {
   MemoryState* state = memoryState;
   RuntimeAssert(!state->gcInProgress, "Recursive GC is disallowed");
 
-  MEMORY_LOG("Garbage collect\n")
+  GC_LOG(">>> GC: toFree %d toRelease %d\n", state->toFree->size(), state->toRelease->size())
 
 #if GC_ERGONOMICS
   auto gcStartTime = konan::getTimeMicros();
 #endif
 
   state->gcInProgress = true;
 
+  incrementStack(state);
```

Comment:
Do I understand correctly that long-living stack references become cycle candidates during every GC?

Reply:
Exactly. Any alternatives?

Reply:
Not yet, but I believe we should eventually apply some efforts to find some.